### ISSUE TITLE
Logout Url : Adding id_token_hint as optionnal parameter

### ISSUE THREAD

**Issue by cedrox** [daggerhart/openid-connect-generic#38]
Hello,
Following openID documentation http://openid.net/specs/openid-connect-session-1_0.html#RPLogout, id_token_hint prevent some denial of service attack and is considered as a best practice. Could you help me to implement it ?
I just wanted to concatenate $url from get_end_session_logout_redirect_url( $redirect_url ) function in \plugins\openid-connect-generic-dev\includes\openid-connect-generic-client-wrapper.php like this :
	$url .= '&id_token_hint=' . $id_token_hint;  
but I don't know how to get back this token properly...

Thanks very much for your work !

**Comment by daggerhart** [comment]
Hi @cedrox,

I'm not 100% sure this is right, but maybe you can test it or otherwise provide feedback.  Here is a patch rolled against the dev branch.

```
diff --git a/includes/openid-connect-generic-client-wrapper.php b/includes/openid-connect-generic-client-wrapper.php
index a2ceeb2..2c3efb9 100644
--- a/includes/openid-connect-generic-client-wrapper.php
+++ b/includes/openid-connect-generic-client-wrapper.php
@@ -261,6 +261,14 @@ class OpenID_Connect_Generic_Client_Wrapper {
 		}
 
 		$url .= 'post_logout_redirect_uri=' . urlencode( $redirect_url );
+
+		// id_token_hint recommended: https://openid.net/specs/openid-connect-session-1_0.html#RPLogout
+		$id_token = get_user_meta( wp_get_current_user()->ID, 'openid-connect-generic-last-id-token-claim', TRUE );
+
+		if ( !empty( $id_token ) ) {
+			$url .= '&id_token_hint=' . base64_encode( wp_json_encode( $id_token ) );
+		}
+
 		return $url;
 	}
 
```

**Comment by cedrox** [comment]
Thanks a lot @daggerhart !
I try it but here the output of my OAuth Server 
error:invalid_client
error_description:Invalid post_logout_redirect_uri

I don't know if there is a kind of process on id_token_hint but the OAut server I work with force the use of it and this is no more optionnal...

Thanks

**Comment by daggerhart** [comment]
I'm not sure, but I think that means the `id_token_hint` is working. But there might be a problem with the `post_logout_redirect_uri`.

Is there somewhere in your IDP where you need to specify allowed redirect uris? Or do you have some documentation about your IDP you can link me to for further investigation?

**Comment by cedrox** [comment]
Unfortunatly, I don't have documentation about how open Id is configured on my IDP. What's the  difference between authorized / return url and post_logout_redirect_url ?
Thanks for your time and feedback !!

**Comment by daggerhart** [comment]
@cedrox, there is a new PR from someone implementing this. Do you have a moment to test it with your IDP? 

#40

**Comment by cedrox** [comment]
Yes, with pleasure ! I will test it and come back to you tomorrow max :)

**Comment by skidos** [comment]
Here some specification about parameters that are passed as query parameters in the logout request:
## id_token_hint
**RECOMMENDED**. Previously issued ID Token passed to the logout endpoint as a hint about the End-User's current authenticated session with the Client. This is used as an indication of the identity of the End-User that the RP is requesting be logged out by the OP. The OP need not be listed as an audience of the ID Token when it is used as an id_token_hint value.

## post_logout_redirect_uri
**OPTIONAL**. URL to which the RP is requesting that the End-User's User Agent be redirected after a logout has been performed. The value MUST have been previously registered with the OP, either using the post_logout_redirect_uris Registration parameter or via another mechanism. If supplied, the OP SHOULD honor this request following the logout.

Except that, the post_logout_redirect_uri doesn't work if the id_token_hint is not provided. The identification of the client and the retrieval of the value of the post_logout_redirect_uri parameter associated with it, are necessary to be able to redirect after the logout.

**Comment by geopal** [comment]
I implemented the changes and I tested it on IdentityServer4 (http://docs.identityserver.io/en/release/index.html), but it cannot bypass the prompt asking for disconnection.

**Comment by daggerhart** [comment]
@geopal Interesting. Did you implement the patch I had pasted in this thread, or PR #40 ?  I'm pretty sure the patch from this thread was wrong, I was just guessing at what needed to be sent as the id_token_hint.

Their documentation says: http://docs.identityserver.io/en/release/endpoints/endsession.html#refendsession

> When the user is redirected to the endpoint, they will be prompted if they really want to sign-out. This prompt can be bypassed by a client sending the original id_token received from authentication. This is passed as a query string parameter called `id_token_hint`.

So hopefully once we are able to send them to correct id_token_hint you will bypass the prompt.

**Comment by geopal** [comment]
@daggerhart I implemented the [#40](https://github.com/daggerhart/openid-connect-generic/pull/40) patch. Output seems to be correct according to the [IdentityServer 4 documentation](http://docs.identityserver.io/en/release/endpoints/endsession.html#refendsession). Using [Fiddler,](http://www.telerik.com/fiddler) I have captured and compared the id_token_hint which our plugin creates, with the id_token_hint of an app which succesfully operates on the exact same Identit yServer. Both clients produce a base64 encoded string split by dots in 3 sections. Although, our plugin seems to fail sending the correct id_token_hint. I am still investigating the issue..

**Comment by cedrox** [comment]
Oups, look like I forget to do a follow up on this. Sorry for that !
I test the changes and it work great for me. 
Let me know if you need some implementation details but it was pretty straight forward.
Thanks

**Comment by geopal** [comment]
I confirm that the patch works fine with identityServer4 (http://docs.identityserver.io/). The client may also send a post_logout_redirect_uri parameter to allow the user to redirect back to the client after sign-out. BE CAREFULL: The value must match (EXACTLY) one of the client’s pre-configured PostLogoutRedirectUris. Be aware of the trailing slashes. I had a problem having https://www.example.com for to PostLogoutRedirectUris and I used https://www.example.com/ (with traling slash) post_logout_redirect_uri. That did not work.

**Comment by cedrox** [comment]
Thanks for those inputs !!

**Comment by daggerhart** [comment]
I've merged #40 into dev branch. Will cleanup some other issues and make a new release soon. Thanks everyone for the help!
